examples/autoignition-mpi.py

```diff
@@ -52,73 +54,115 @@
 from mirgecom.boundary import AdiabaticSlipBoundary
 from mirgecom.initializers import MixtureInitializer
 from mirgecom.eos import PyrometheusMixture
+
+from mirgecom.logging_quantities import (
+    initialize_logmgr,
+    logmgr_add_many_discretization_quantities,
+    logmgr_add_device_name,
+    logmgr_add_device_memory_usage,
+    set_sim_state
+)
+
 import cantera
 import pyrometheus as pyro
 
 logger = logging.getLogger(__name__)
 
 
+class MyError(Exception):
+    """Simple exception to kill the simulation."""
+
+    pass
+
+
+class HealthCheckError(MyError):
+    """Simple exception to indicate a health check error."""
+
+    pass
```

Comment:
@MTCam Since we're back to exceptions, should we add the previous simulation exceptions back in?

Reply:
I suggest we make a new one, and maybe as a follow-on PR.  The old one included exceptions we don't use, and these are far simpler.  What would we think about adding only this one as a library-provided exception:

```python
class FatalSimulationError(Exception):
    pass
```
Potentially we could beef it up with a message or so.

Then, users could inherit this for their own purposes, for stuff like:
```python
class MyHealthCheckError(FatalSimulationError):
    pass
```

which would appear in the driver.

Reply:
@MTCam The only reason I mention this here is that I don't like drivers creating/defining new exception classes. It feels weird/wrong. I'd rather we just import them from some pre-defined module or just use built-in python exceptions. @majosm what do you think?

Reply:
Right, then it seems silly that we would offer an exception from the library that doesn't add anything to the behavior of the built-in `Exception`.  I think these drivers will work the same if we eliminate these exception classes and just `raise` instead.

Reply:
@thomasgibson Please have a look at autoignition to see what it looks like with all built-in exceptions.  There is a dealie in place to force a health-check error at step 5.

Reply:
Ooh, hang on -- we can't do that. We don't want to catch all exceptions in that block (since they may not be synchronized). We only want the ones that the user knows are safe to catch.

Reply:
As for defining exceptions in drivers; I think it's ok -- driver developers should be able to create exceptions that reflect the errors that occur in their callbacks. We can do a common base class, but I'm not sure if it makes sense right now since the base class doesn't define any interface. Maybe we should punt on it for now and see how things look once the dust has settled a bit?

Reply:
> > @thomasgibson Please have a look at autoignition to see what it looks like with all built-in exceptions. There is a dealie in place to force a health-check error at step 5.
> 
> Ooh, hang on -- we can't do that. We don't want to catch all exceptions in that block (since they may not be synchronized). We only want the ones that the user knows are safe to catch.

Thank you, nice catch.  These are exactly the subtle issues that I am worried about running afoul of.

Reply:
@majosm Can you please have a fresh look at `autoignition-mpi.py` with this in-mind?  The thing I have there now may be acceptable and help address @thomasgibson's concern/preference.

Reply:
> As for defining exceptions in drivers; I think it's ok -- driver developers should be able to create exceptions that reflect the errors that occur in their callbacks. We can do a common base class, but I'm not sure if it makes sense right now since the base class doesn't define any interface. Maybe we should punt on it for now and see how things look once the dust has settled a bit?

I concur with this - but also in trying to address @thomasgibson's preference, I do like the message that `RuntimeError` offers.  I do get a log message with the program output, but it is "nice" to see the health check indication at the end of the exception spew as well.

Reply:
> @majosm Can you please have a fresh look at `autoignition-mpi.py` with this in-mind? The thing I have there now may be acceptable and help address @thomasgibson's concern/preference.

So, when you do `except <exceptiontype>:`, you're telling Python that you want to catch any error that inherits from `<exceptiontype>` and do something (in this case write vis/restart). Other exceptions (those that don't inherit from `<exceptiontype>`) will keep bubbling up the call stack either to be handled elsewhere or to terminate the program if nothing handles them. (The `raise` with no arguments just means that you want to let the same exception keep bubbling up after you've done stuff.)

When `<exceptiontype>` is `BaseException`, it will catch every possible exception. We don't want to do that here since there's lots of code inside this block that could potentially raise some kind of exception, and most of it is unaware that it's in a distributed environment. We want to restrict the handling to exceptions that the driver developer knows are globally synchronized, hence creating a base class in the driver. (We could give the base class a more descriptive name indicating as such, if it seems necessary; `GlobalError`? `MyGlobalError`?)

Reply:
Thanks for those details @majosm - I understand that the current code is about to blow my leg off if some component I'm using throws anything at all (this explains some of the linting warnings i was seeing!).  How would we feel about:

```python
class MyRuntimeError(RuntimeError):
    pass
```

in the driver and then `except MyRuntimeError`?  I do see the draw of using an additional `MyHealthCheckError(MyRuntimeError)`, but since `RuntimeError` offers a built-in message, maybe we don't really need a health-specific one?

Reply:
Sounds good to me. In some cases it may be useful to differentiate the types so you can handle each type differently, but for these cases I think the same type with different messages should be fine.

Reply:
btw, a million thanks to both of you for your patience and attention on this @majosm and @thomasgibson.  I feel a little silly for dragging my feet so much about the exceptions in-particular.  I agree that the exceptions have improved the error handling situation by a lot.